zerver/decorator.py

```diff
@@ -114,6 +114,11 @@ def wrapper(request, user_profile, *args, **kwargs):
         # type: (HttpRequest, UserProfile, *Any, **Any) -> HttpResponse
         if not user_profile.is_realm_admin:
             raise JsonableError(_("Must be a realm administrator"))
+
+        if not check_subdomain(get_subdomain(request), user_profile.realm.subdomain):
+            logging.warning("User %s attempted to access wrong subdomain %s" % (
+                user_profile.email, get_subdomain(request)))
+            raise JsonableError(_("Account is not associated with this subdomain"))
```

Comment:
Take look at `rest_dispatch` -- that is the function we have declared in our views files for the views that only directly declare `has_request_variables`.  `rest_dispatch` actually wraps the relevant actual view function with either `authenticated_json_view` or `authenticated_rest_api_view`.  

So, the security invariant for our authenticated views is that they all either use `rest_dispatch` or have a direct authentication decorator on them.  If a view function didn't have an authentication decorator on it, it wouldn't work at all, since the `user_profile` argument is passed into our views by the decorators.  Does that make sense?